Objective  
Load, inspect, clean, and prepare a structured retail dataset using Python and Pandas. This is the
bread and butter of any data role — you will do this hundreds of times in your career, so build the
muscle memory now

In [39]:
import pandas as pd
df = pd.read_csv('SampleSuperstore.csv', encoding='latin-1')
df.head(10)
# pd.read_csv() — reads a CSV file into a DataFrame

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
5,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
6,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800,4,0.00,1.9656
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
8,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040,3,0.20,5.7825
9,10,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000,5,0.00,34.4700


In [40]:
# Examine the dataset’s shape, data types, missing values, and summary statistics
# Step 1: Check the shape:

print("Shape:", df.shape) # (rows, columns)

Shape: (9994, 21)


In [41]:
# Step 2: Run .info() to see data types and non-null counts:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [42]:
# Step 3: Run .describe() for summary statistics of numerical columns:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [43]:
# Step 3: Run .describe() for summary statistics of numerical columns:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [44]:
#Check for missing values:
print(df.isnull().sum())

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64


The dataset has 9,994 rows and 21 columns covering order, customer, product,
and financial data. There are no missing values in any column — every field
is fully populated. The Profit column shows a wide range (min: -6,599 to
max: 8,399), suggesting some orders incur significant losses, which is worth
investigating later.



Task 3: Handle Missing Values  
What to do: Fill missing values in numerical columns with median and categorical columns with
mode.  
How to do it:  
• Step 1: Identify which columns have missing values (from Task 2).  
• Step 2: For numerical columns, fill with median (more robust than mean if outliers exist):

In [45]:
# Step 1: Identify which columns have missing values (from Task 2).
#• Step 2: For numerical columns, fill with median (more robust than mean if outliers exist):
# Example for a numerical column
df['Sales'].fillna(df['Sales'].median(), inplace=True)

/tmp/ipykernel_4382/1288685314.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Sales'].fillna(df['Sales'].median(), inplace=True)


In [46]:
# Step 3: For categorical columns, fill with mode (most frequent value):
# Example for a categorical column
df['Ship Mode'].fillna(df['Ship Mode'].mode()[0], inplace=True)

/tmp/ipykernel_4382/1604052869.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Ship Mode'].fillna(df['Ship Mode'].mode()[0], inplace=True)


In [47]:
# Step 4: Verify no missing values remain:
print(df.isnull().sum())

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64


After checking with df.isnull().sum(), the dataset has 0 missing values across
all 21 columns. No filling was needed. Had there been missing values, I would
have used median for numerical columns (robust to outliers) and mode for
categorical columns (most frequent value).

Task 4: Remove Duplicates  
What to do: Check for and remove duplicate rows.  
How to do it:

In [48]:
# Step 1: Count duplicates:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [49]:
# Step 2: Remove them:
df = df.drop_duplicates()
print("Shape after removing duplicates:", df.shape)

Shape after removing duplicates: (9994, 21)


The dataset contained 0 duplicate rows, so no rows were removed. The shape
remains (9994, 21). df.drop_duplicates() was still applied as a best practice
to ensure data integrity going forward.

Task 5: Convert Data Types  
What to do: Convert Order Date and Ship Date columns to proper datetime format.  
How to do it:

In [50]:
# Step 1: Convert the date columns:
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])
# Tool: pd.to_datetime() — converts strings to datetime objects

In [51]:
#Step 2: Verify the conversion:
print(df.dtypes)

Row ID                    int64
Order ID                 object
Order Date       datetime64[ns]
Ship Date        datetime64[ns]
Ship Mode                object
Customer ID              object
Customer Name            object
Segment                  object
Country                  object
City                     object
State                    object
Postal Code               int64
Region                   object
Product ID               object
Category                 object
Sub-Category             object
Product Name             object
Sales                   float64
Quantity                  int64
Discount                float64
Profit                  float64
dtype: object


In [52]:
# Step 3: Spot-check a few values to make sure dates look correct:
df[['Order Date', 'Ship Date']].head()
# Tool: pd.to_datetime() — converts strings to datetime objects

,Order Date,Ship Date
0,2016-11-08,2016-11-11
1,2016-11-08,2016-11-11
2,2016-06-12,2016-06-16
3,2015-10-11,2015-10-18
4,2015-10-11,2015-10-18


Task 6: Create Derived Features  
What to do: Create three customer-level summary metrics: total spending, order frequency, and
average order value.  
How to do it:

In [53]:
# Step 1: Calculate total spending per customer:
total_spending = df.groupby('Customer ID')['Sales'].sum().reset_index()
total_spending.columns = ['Customer ID', 'total_spending']

In [54]:
# Step 2: Calculate order frequency (number of orders per customer):
order_freq = df.groupby('Customer ID')['Order ID'].nunique().reset_index()
order_freq.columns = ['Customer ID', 'order_frequency']

In [55]:
# Step 3: Calculate average order value:
avg_order = df.groupby('Customer ID')['Sales'].mean().reset_index()
avg_order.columns = ['Customer ID', 'avg_order_value']

In [56]:
# Step 4: Merge all into a customer summary table:
customer_summary = total_spending.merge(order_freq, on='Customer ID').merge(avg_order, on='Customer ID')
customer_summary.head(10)

,Customer ID,total_spending,order_frequency,avg_order_value
0,AA-10315,5563.560,5,505.778182
1,AA-10375,1056.390,9,70.426000
2,AA-10480,1790.512,4,149.209333
3,AA-10645,5086.935,6,282.607500
4,AB-10015,886.156,3,147.692667
5,AB-10060,7755.620,8,430.867778
6,AB-10105,14473.571,10,723.678550
7,AB-10150,966.710,5,80.559167
8,AB-10165,1113.838,8,79.559857
9,AB-10255,914.532,9,65.323714


In [57]:
# Step 5: Save the cleaned DataFrame:
df.to_csv('superstore_cleaned.csv', index=False)
customer_summary.to_csv('customer_summary.csv', index=False)

# Tool: df.groupby(), .sum(), .nunique(), .mean(), .merge(), .to_csv()